### ISSUE TITLE
When dashing audio file with mp4box, drift occurs between MPD timeline and tfdt for no reason

### ISSUE THREAD

**Issue by sandersaares** [gpac/gpac#949]
Command: `mp4box -dash-strict 4000 -out d:\tmp\Manifest.mpd -profile "dashavc264:live" A0-24h_TestPattern.mp4`
Version: latest nightly win64
Input file: https://media.axprod.net/PublicVideos/mp4box_drift/A0-24h_TestPattern.mp4

Expected result: dashing success, no unexpected timing problems.

Actual result: errors related to timing are printed in output

```
[...]
[DASH] Segment d:/tmp/A0-24h_TestPattern_dash189.m4s (Number #189): drift between MPD timeline and tfdt exceeds 50% of segment duraion (MPD time minus TFDT 2.005333 secs) - bitstream will not be compliant, try using segment template or reencode
[DASH] Segment d:/tmp/A0-24h_TestPattern_dash190.m4s (Number #190): drift between MPD timeline and tfdt exceeds 50% of segment duraion (MPD time minus TFDT 2.016000 secs) - bitstream will not be compliant, try using segment template or reencode
[DASH] Segment d:/tmp/A0-24h_TestPattern_dash191.m4s (Number #191): drift between MPD timeline and tfdt exceeds 50% of segment duraion (MPD time minus TFDT 2.026667 secs) - bitstream will not be compliant, try using segment template or reencode
[DASH] Segment d:/tmp/A0-24h_TestPattern_dash192.m4s (Number #192): drift between MPD timeline and tfdt exceeds 50% of segment duraion (MPD time minus TFDT 2.037333 secs) - bitstream will not be compliant, try using segment template or reencode
[... continues for every segment]
```

Same if I use `-dash` instead of `-dash-strict`.

**Comment by rbouqueau** [comment]
Hi Sander,

Thanks for reporting. Actually this (annoying) message is wrong but the content should be ok (let me know if otherwise).

@jeanlf 68f1c24c469dfb5b5c9d73f6cc1269f6d0fd0512 Do you remember why this commit introduced the block that emits this message? The computation is based on ```MaxSegmentDuration``` instead of the average duration.

**Comment by sandersaares** [comment]
Content does appear to be incorrect.

For example, segment 1000 is expected to start at 4x1000=4000 seconds. Instead, it contains content starting from 3985 seconds which appears to match the error messages.

**Comment by rbouqueau** [comment]
Ok then the issue is older than the commit I cited. I let @jeanlf see if he remembers about these changes.

**Comment by ZhangkaiKevin** [comment]
I have the same issue

**Comment by rayfoss** [comment]
also same issue... Running the current build https://github.com/gpac/gpac/commit/a7a79f459a76074f1808dd168e0f67d41eec8759

Ubuntu. 17.10

**Comment by aswin1980** [comment]
Hi,
I'm having the similar issue. After 30 Minutes I'm seeing these messages

[DASH] Segment /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/vo/liveme4554.m4s (Number #4554): drift between MPD timeline and tfdt exceeds 50% of segment duraion (MPD time minus TFDT 2091.700000 secs) - bitstream will not be compliant, try using segment template or reencode


The below commands are used to convert the mp4 file. The mp4file is created using openrtsp. For 10 minutes everything seems to be working in the browser after that it degrades and stops.

ffmpeg -hide_banner -loglevel panic -i /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/1.mp4 -vcodec copy -an -y /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/live.mp4

    /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/build/bin/MP4Box -dash-strict 1000 -dash-ctx /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/vo/dash-live.txt -profile live -subdur 10000 -segment-name liveme -mpd-refresh 20 -insert-utc -time-shift 60 -bs-switching no -dynamic -out /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/vo/manifest /mnt/35af65e4-e4c5-496f-899d-2f72e3c24cb8/gpac/gpac/live.mp4:#video -frag 1000

Any idea how to solve it.

Thanks
Aswin

**Comment by aureliendavid** [comment]
@aswin1980 : could you upload your test video somewhere? (e.g. https://www.mediafire.com/filedrop/filedrop_hosted.php?drop=eec9e058a9486fe4e99c33021481d9e1826ca9dbc242a6cfaab0fe95da5e5d95)

I think we might have a solution for the audio case, but need to do more tests for video.

**Comment by rbouqueau** [comment]
@aureliendavid FYI I could easily reproduce with the content in the first post.

**Comment by aureliendavid** [comment]
yep that's the audio-only case for which a solution might be to allow segments to be slighly longer than the dash duration (which apparently is acceptable?) to catch up with the media time and avoid drift (which seem to work for the OP's file)

but for video there already is a similar "trick" to adapt segment duration to split on keyframes, but it still allows drift because it doesn't check media time - I need to do more test on this

**Comment by sandersaares** [comment]
My preferred segmentation algorithm is to ensure that the media sample matching the segment start timestamp is within that segment. DASH and DASH-IF IOP allow for some flexibility here but this seems to be most foolproof algorithm.

At the moment, due to this issue, we had to move some of our workflows to a different packager that implements such an algorithm and this segmentation logic (a segment starts with the sample that contains the segment start) is proving very reliable and robust.

**Comment by aureliendavid** [comment]
@sandersaares : are you using both audio and video?

for audio, that algorithm you describe is what the solution I talked about above will end up implementing (i.e.: even if the segment is already longer than the dash duration we keep adding samples until we reach the media time of the segment boundary) 

this is possible because with audio we can split anywhere (all samples are random access points)

however, for video, we need to split on keyframes - and the interval between keyframes might not be constant, and might not be a divisor of the dash duration, for this kind of media we need to have some criteria on which SAP do we split on 

even then I think some drift is inevitable in some cases, the important thing is that it stays low and doesn't keep growing like in the OP's case

now there's also the case where the dash duration asked is simply too low compared to the keyframe intervals, in that case I don't know if we can do anything

for reference, a way to check the timestamps of the keyframes : 

`ffprobe -loglevel error -select_streams v:0 -show_entries frame=key_frame,pkt_pts_time -of csv=print_section=0 input.mp4 | awk -F',' '/^1,/{print $2}'`



_(this is my understanding of the whole thing but I don't know dash very well so feel free to correct me if I'm wrong)_

**Comment by sandersaares** [comment]
Yes, we are using this algorithm for both audio and video.

Yes, if keyframes are not well aligned with segments then segments will not be well aligned with their boundaries in DASH. This is not avoidable without making keyframe distance more DASH-suitable and it is fine, really. What using the abovementioned algorithm does is it ensures that each segment starts at least close to its expected start position and always includes the sample at the segment start position, which seems to be good enough to cover all our bases.

For cases where the segment starts on a non-keyframe, one would have to extend it back in time to cover the most recent keyframe, of course, as you say.

I agree with your conclusion that if keyframe interval is beyond segment duration, it is hopeless.

**Comment by aureliendavid** [comment]
so the global rule would be something like: 

`segment_start = max( RAP <= segment_number*dash_duration )`

?

if every sample is a RAP (audio) then it'll be the sample containing the boundary, otherwise it'll be the previous keyframe (even if it is further than the next keyframe over the boundary)

that seems doable, I'll look into it

**Comment by sandersaares** [comment]
Yes, that seems to be an accurate representation of the algorithm I had in mind.

**Comment by rbouqueau** [comment]
> yep that's the audio-only case for which a solution might be to allow segments to be slighly longer than the dash duration (which apparently is acceptable?) to catch up with the media time and avoid drift (which seem to work for the OP's file)

```-dash``` is suppose to behave this way (with segment duration oscillating around the requested duratin). ```-dash-strict``` would fix the duration (inf. or eq. for audio). If this behaviour changed, that's a regression.

> ffprobe -loglevel error -select_streams v:0 -show_entries frame=key_frame,pkt_pts_time -of csv=print_section=0 input.mp4 | awk -F',' '/^1,/{print $2}'

Also http://gpac.io/2015/11/02/check-key-frame-alignment-with-mp4box/ :)

**Comment by aureliendavid** [comment]
> -dash is suppose to behave this way (with segment duration oscillating around the requested duratin). -dash-strict would fix the duration (inf. or eq. for audio). If this behaviour changed, that's a regression.

as far as I can tell, in the code, there are currently no difference between -dash and -dash-strict

I think this is the change that broke it: https://github.com/gpac/gpac/commit/4b1f24c18ca3d9c17bf5b5b859646a4bbf8e22d6#diff-820b3363494bfb7b10cec0e878405badR1679

this makes audio always behave as -dash-strict

**Comment by aureliendavid** [comment]
After discussion, here's what we're thinking to fix the issues mentioned here: 

 1. mark `-dash-strict` as deprecated
 2. fix audio drift with `-dash` (to fix the OP's issue)
 3. for video, keep the current behavior by default for compatibility (i.e.: allow segments to grow up to 150% of the dash duration without checking the boundaries)
 4. add an option to use @sandersaares's algorithm for video (something like `-bound`)

If there are no objections to this plan, I'll go ahead with it.

**Comment by aureliendavid** [comment]
The changes mentioned above are now merge into master. 

If the people who were having issues with drift can retest with the latest version and tell us if it's better now, it would be appreciated.

**Comment by DanielMartine** [comment]
HI @aureliendavid thanks for your work, examining  GPAC version 0.7.2-DEV-rev386-gc66782d8-master i can't find additional attributes such -bound as you say, is this the correct upgreded version? 
Which os are compatible with?

**Comment by DanielMartine** [comment]
Hi @aureliendavid  thanks for your work! I see that in this post there are no examples with -live-dash, May i ask to be sure if your update work also for live use?
Thanks!

**Comment by DanielMartine** [comment]
after some checks I can see that the stability is noticeably degraded with or without -bound, I still have to be sure, but the only change to the setup was the update, tomorrow i will check with a previous version to be sure

**Comment by aureliendavid** [comment]
@DanielMartine: let's take the problems one by one - do you get the 404s with and without -bound? what player are you using?

thanks

**Comment by DanielMartine** [comment]
Hi! without bound the streaming is even more unstable, lasts a couple of minutes at most and always crashes with 404 error and whitout MP4Box edvise. (also before the timestamp discrepancy)
with bound the streaming seems to work, but after the end of the first loop it becomes impossible to update the page due to 404 error, but if I do not update the page the problem does not occur and I can continue to view the stream.

**Comment by aureliendavid** [comment]
let's forget about -bound for now

can you confirm that without it, it works with revision 66f0361bd1759c174e9c65ef7cfad57154c9bd0e and not with revision c0e1ec6f2cb1d7eac1cc6b45dd29dea5a0beaaa3 ?

**Comment by DanielMartine** [comment]
Ok, i will be here within an hour with the tests you requested, may i ask you to decrease the time needed how to get the previous version via git instead `git clone https://github.com/gpac/gpac.git`

**Comment by DanielMartine** [comment]
letest version without using -bound on ubuntu server 16.04.3 with dash.js contrib dash and also only with dash.js give to me only few minutes (1 or 2) of live streaming before it stop due to 404 on chrome consolle

**Comment by DanielMartine** [comment]
But i need a help to restore a previous version, serching i still not found how to

**Comment by aureliendavid** [comment]
you can do 

```
git clone https://github.com/gpac/gpac.git
cd gpac
git checkout 66f0361bd1759c174e9c65ef7cfad57154c9bd0e
```

and build from there

I may have been overconfident with multiple changes in one commit. I'll revert it if I don't find what's wrong with it by tomorrow and reintroduce the changes one by one. I need to find a reliably reproducible test case to see what's wrong though.

**Comment by DanielMartine** [comment]
With 66f0361 seems to work correctly with video.js contrib dash and dash.js, no 404.
the reason why I arrived on this page is because after 52 hours of testing I finally got the infamous error drift beetween MPD.. I would like to get a stable process.
Again thanks for your work feel free to ask me all the tests I can produce.

**Comment by aureliendavid** [comment]
Ok I think I found what I did wrong in my initial commit, causing the 404s (with and without -bound). 

The latest version should now work at least as well as the old ones for video (which is not perfect but that's another subject), and should fix the audio drift issue from the original post. 

If this is confirmed, I suggest closing this issue. If there are other problems (with -bound, buffering, etc.), please open other issues so as to keep discussion easy to follow. 

@DanielMartine: can you confirm that the latest version works the same as https://github.com/gpac/gpac/commit/66f0361bd1759c174e9c65ef7cfad57154c9bd0e for your test case?

@sandersaares: can you confirm that the original issue about audio drift is now fixed?

Thanks all for you help. 


(The -bound option should also work properly but let's not dwell on that here).

**Comment by DanielMartine** [comment]
last revision seems to work fine with `-dash-live `+ `-bound` 
to be sure I will run mp4box for 3 - 4 days hoping it will go all right.
I will send here the result.
Hoping for close the issue!
Thanks!

**Comment by DanielMartine** [comment]
here are more then 72 hours still audio-video perfect sync, whithout any error.

**Comment by DanielMartine** [comment]
Hi @aureliendavid I noticed a problem, I do not know if it is appropriate to talk about it here or if you prefer that I open a new issue: with `-dash-live` + `-bound` if I watch for a long time video and audio go out of sync (only after some repetitons), without any warning or error, if I reload the page video and the audio returns synchronized, it would seem that `-bound` has no effect while someone is watching?!

**Comment by aureliendavid** [comment]
Please open a new issue for this, with as much details as possible about the test case and the result. 

I'm closing this one since the original issue looks resolved.

**Comment by sandersaares** [comment]
I confirm the original issue in the OP is solved and `-bound` appears to be behaving according to expectations. Thanks for taking care of it!
